(PUP-9579) Use device specific ssl_context

joshcooper opened the pull request:
Previously the device application failed when trying to download a
catalog for a device, because the SSL statemachine doesn't take a
`certname` keyword argument.

Also, we have to push the device specific `ssl_context` onto the context
before making network connections on behalf of the device. Otherwise, we
end up using the node's normal ssl cert.

Also don't pluginsync twice per device.

joshcooper commented:
This needs to go in 6.4.x

puppetcla commented:
CLA signed by all contributors.

kris-bosland commented on spec/unit/application/device_spec.rb:
```diff
@@ -75,30 +75,39 @@
     end
 
     it "should set waitforcert to 0 with --onetime and if --waitforcert wasn't given" do
+      Puppet[:certname] = 'device.example.com'
```
Would this be better in a describe block with a begin block?

joshcooper replied:
Updated

melissa commented on lib/puppet/application/device.rb:
```diff
@@ -372,9 +377,9 @@ def find_resources(type, name)
-  def setup_host(name)
+  def setup_host
     waitforcert = options[:waitforcert] || (Puppet[:onetime] ? 0 : Puppet[:waitforcert])
-    sm = Puppet::SSL::StateMachine.new(certname: name, waitforcert: waitforcert)
```
https://github.com/puppetlabs/puppet/blob/85be99da5bb4cdd011ee68b00746ec2e9b6e65fc/lib/puppet/ssl/state_machine.rb#L210-L213
It looks like nothing was ever done with `certname` in the statemachine

This comment is mostly for my peace of mind, knowing nothing will be impacted by not including this var in this method call

joshcooper replied:
An earlier version of the client state machine accepted a `:certname` keyword argument, but I took it out, and forgot to fix `device.rb`.

melissa commented on lib/puppet/application/device.rb:
```diff
@@ -319,15 +327,12 @@ def main
             end
           else
```
if `options[:libdir]`, we'll enter this block, but we won't have gone through the above block first. Will we still have `ssl_context` defined in that case? My gut says no. Also, will `ssl_context` be available outside of the block that starts on line 272?

joshcooper replied:
Good point. @DavidS if `--libdir` is specified, then I assume we want to skip pluginsync, but we do want to apply the configuration to the target devices (in which case we'll need to perform ssl bootstrap earlier)?

joshcooper replied:
I added another commit that matches my comment above.

melissa commented on lib/puppet/application/device.rb:
```diff
@@ -319,15 +327,12 @@ def main
-            configurer = Puppet::Configurer.new
-            configurer.run(:network_device => true, :pluginsync => Puppet::Configurer.should_pluginsync? && !options[:libdir])
+            Puppet.override(ssl_context: ssl_context) do
+              require 'puppet/configurer'
```
Why do we need this `require` here and we don't need a similar `require` before line 280?

joshcooper replied:
Good point. The device application already requires it on line 2, so it's not needed here at all. I'll remove it.

joshcooper commented:
Need to retarget 6.4.x and update mocha/rspec tests

joshcooper commented:
Verified we (for each device) get a client cert, pluginsync and apply a catalog:

```
$ bx puppet device --deviceconfig device.conf -v -w 4 --server puppet.delivery.puppetlabs.net
Info: Downloaded certificate for ca from puppet.delivery.puppetlabs.net
Info: Downloaded certificate revocation list for ca from puppet.delivery.puppetlabs.net
Info: Creating a new SSL key for nexus1
Info: csr_attributes file loading from /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/csr_attributes.yaml
Info: Creating a new SSL certificate request for nexus1
Info: Certificate Request fingerprint (SHA256): 68:BB:8D:84:F6:F9:20:D6:BF:DC:3B:7B:72:FE:29:C5:D1:F4:0D:31:C3:13:3F:19:57:61:8B:B9:5A:40:E4:C0
Info: Downloaded certificate for nexus1 from puppet.delivery.puppetlabs.net
Info: Retrieving pluginfacts
Info: Retrieving plugin
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/provider]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/provider/local_security_policy]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/provider/local_security_policy/policy.rb]/ensure: defined content as '{md5}eee8076ccc563dd96e9cecfbecea7694'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/provider/registry_key]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/provider/registry_key/registry.rb]/ensure: defined content as '{md5}57d97e6cf5e20370af6caff63f575b6a'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/provider/registry_value]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/provider/registry_value/registry.rb]/ensure: defined content as '{md5}70f825ee0a196ea0e3030bafd979aec6'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/type]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/type/local_security_policy.rb]/ensure: defined content as '{md5}66875e9cc349beb0edf5ca3820f04373'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/type/registry_key.rb]/ensure: defined content as '{md5}833cac29fc9cbc761dfb1616d3c7cc8b'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet/type/registry_value.rb]/ensure: defined content as '{md5}9e7cafee6cc3a514e05b9a34c2e43292'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet_x]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet_x/lsp]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet_x/lsp/security_policy.rb]/ensure: defined content as '{md5}74c4dd4da3a3328fe698f363cef66823'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet_x/puppetlabs]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet_x/puppetlabs/registry.rb]/ensure: defined content as '{md5}1132a6f66c4e7d549ac90f9f58bd73e2'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet_x/twp]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/lib/puppet_x/twp/inifile.rb]/ensure: defined content as '{md5}d9e1e62229899e191836fbc5254555d2'
Info: Retrieving locales
Info: starting applying configuration to nexus1 at file:///Users/josh/work/puppet/nexus1.conf
Info: Using configured environment 'production'
Info: Caching catalog for nexus1
Info: Applying configuration version '1553890940'
Notice: foo
Notice: /Stage[main]/Main/Node[default]/Notify[foo]/message: defined 'message' as 'foo'
Info: Node[default]: Unscheduling all events on Node[default]
Info: Creating state file /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/state/state.yaml
Notice: Applied catalog in 0.01 seconds
Info: Downloaded certificate for ca from puppet.delivery.puppetlabs.net
Info: Downloaded certificate revocation list for ca from puppet.delivery.puppetlabs.net
Info: Creating a new SSL key for nexus2
Info: csr_attributes file loading from /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/csr_attributes.yaml
Info: Creating a new SSL certificate request for nexus2
Info: Certificate Request fingerprint (SHA256): A3:2D:7B:80:4B:9C:02:15:A9:9E:BC:C6:E9:18:E4:F3:7D:F6:48:06:91:A5:AC:C3:3A:C5:83:F0:0D:A8:AF:DA
Info: Downloaded certificate for nexus2 from puppet.delivery.puppetlabs.net
Info: Retrieving pluginfacts
Info: Retrieving plugin
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/provider]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/provider/local_security_policy]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/provider/local_security_policy/policy.rb]/ensure: defined content as '{md5}eee8076ccc563dd96e9cecfbecea7694'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/provider/registry_key]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/provider/registry_key/registry.rb]/ensure: defined content as '{md5}57d97e6cf5e20370af6caff63f575b6a'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/provider/registry_value]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/provider/registry_value/registry.rb]/ensure: defined content as '{md5}70f825ee0a196ea0e3030bafd979aec6'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/type]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/type/local_security_policy.rb]/ensure: defined content as '{md5}66875e9cc349beb0edf5ca3820f04373'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/type/registry_key.rb]/ensure: defined content as '{md5}833cac29fc9cbc761dfb1616d3c7cc8b'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet/type/registry_value.rb]/ensure: defined content as '{md5}9e7cafee6cc3a514e05b9a34c2e43292'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet_x]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet_x/lsp]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet_x/lsp/security_policy.rb]/ensure: defined content as '{md5}74c4dd4da3a3328fe698f363cef66823'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet_x/puppetlabs]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet_x/puppetlabs/registry.rb]/ensure: defined content as '{md5}1132a6f66c4e7d549ac90f9f58bd73e2'
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet_x/twp]/ensure: created
Notice: /File[/Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/lib/puppet_x/twp/inifile.rb]/ensure: defined content as '{md5}d9e1e62229899e191836fbc5254555d2'
Info: Retrieving locales
Info: starting applying configuration to nexus2 at file:///Users/josh/work/puppet/nexus2.conf
Info: Using configured environment 'production'
Info: Caching catalog for nexus2
Info: Applying configuration version '1553890950'
Notice: foo
Notice: /Stage[main]/Main/Node[default]/Notify[foo]/message: defined 'message' as 'foo'
Info: Node[default]: Unscheduling all events on Node[default]
Info: Creating state file /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/state/state.yaml
Notice: Applied catalog in 0.01 seconds
```
Rerunning the same command, reuses the existing client certs, but does pluginsync and applies catalogs:

```
$ bx puppet device --deviceconfig device.conf -v -w 4 --server puppet.delivery.puppetlabs.net
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: starting applying configuration to nexus1 at file:///Users/josh/work/puppet/nexus1.conf
Info: Using configured environment 'production'
Info: Caching catalog for nexus1
Info: Applying configuration version '1553890963'
Notice: foo
Notice: /Stage[main]/Main/Node[default]/Notify[foo]/message: defined 'message' as 'foo'
Info: Node[default]: Unscheduling all events on Node[default]
Notice: Applied catalog in 0.01 seconds
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: starting applying configuration to nexus2 at file:///Users/josh/work/puppet/nexus2.conf
Info: Using configured environment 'production'
Info: Caching catalog for nexus2
Info: Applying configuration version '1553890968'
Notice: foo
Notice: /Stage[main]/Main/Node[default]/Notify[foo]/message: defined 'message' as 'foo'
Info: Node[default]: Unscheduling all events on Node[default]
Notice: Applied catalog in 0.02 seconds
```

Using `--libdir` skips pluginsync, and uses existing client certs:

```
$ bx puppet device --deviceconfig device.conf -v -w 4 --server puppet.delivery.puppetlabs.net --libdir ~/.puppetlabs/opt/puppet/cache/devices/nexus1/lib
Info: starting applying configuration to nexus1 at file:///Users/josh/work/puppet/nexus1.conf
Info: Using configured environment 'production'
Info: Caching catalog for nexus1
Info: Applying configuration version '1553890994'
Notice: foo
Notice: /Stage[main]/Main/Node[default]/Notify[foo]/message: defined 'message' as 'foo'
Info: Node[default]: Unscheduling all events on Node[default]
Notice: Applied catalog in 0.02 seconds
Info: starting applying configuration to nexus2 at file:///Users/josh/work/puppet/nexus2.conf
Info: Using configured environment 'production'
Info: Caching catalog for nexus2
Info: Applying configuration version '1553891000'
Notice: foo
Notice: /Stage[main]/Main/Node[default]/Notify[foo]/message: defined 'message' as 'foo'
Info: Node[default]: Unscheduling all events on Node[default]
Notice: Applied catalog in 0.02 seconds
```

Cleaning device certs and rerunning, skips pluginsync, but bootstraps new client certs:

```
$ bx puppet ssl clean --target nexus1
Notice: Removed private key /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/ssl/private_keys/nexus1.pem
Notice: Removed certificate /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/ssl/certs/nexus1.pem
$ bx puppet ssl clean --target nexus2
Notice: Removed private key /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/ssl/private_keys/nexus2.pem
Notice: Removed certificate /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/ssl/certs/nexus2.pem
$ bx puppet device --deviceconfig device.conf -v -w 4 --server puppet.delivery.puppetlabs.net --libdir ~/.puppetlabs/opt/puppet/cache/devices/nexus1/lib
Info: Creating a new SSL key for nexus1
Info: csr_attributes file loading from /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus1/csr_attributes.yaml
Info: Creating a new SSL certificate request for nexus1
Info: Certificate Request fingerprint (SHA256): D3:43:AB:8D:EA:23:84:F4:9B:C4:3C:49:DF:F0:9B:81:21:18:C9:69:DE:25:D4:7D:DC:59:9B:61:21:5F:3D:A8
Info: Downloaded certificate for nexus1 from puppet.delivery.puppetlabs.net
Info: starting applying configuration to nexus1 at file:///Users/josh/work/puppet/nexus1.conf
Info: Using configured environment 'production'
Info: Caching catalog for nexus1
Info: Applying configuration version '1553891028'
Notice: foo
Notice: /Stage[main]/Main/Node[default]/Notify[foo]/message: defined 'message' as 'foo'
Info: Node[default]: Unscheduling all events on Node[default]
Notice: Applied catalog in 0.02 seconds
Info: Creating a new SSL key for nexus2
Info: csr_attributes file loading from /Users/josh/.puppetlabs/opt/puppet/cache/devices/nexus2/csr_attributes.yaml
Info: Creating a new SSL certificate request for nexus2
Info: Certificate Request fingerprint (SHA256): 71:69:7F:E4:2C:93:C5:C3:D2:1A:8B:9E:92:2D:EC:04:38:16:04:89:47:91:1B:73:43:08:13:36:AE:97:7C:67
Info: Downloaded certificate for nexus2 from puppet.delivery.puppetlabs.net
Info: starting applying configuration to nexus2 at file:///Users/josh/work/puppet/nexus2.conf
Info: Using configured environment 'production'
Info: Caching catalog for nexus2
Info: Applying configuration version '1553891035'
Notice: foo
Notice: /Stage[main]/Main/Node[default]/Notify[foo]/message: defined 'message' as 'foo'
Info: Node[default]: Unscheduling all events on Node[default]
Notice: Applied catalog in 0.02 seconds
```

DavidS commented:
This will be conflicting behind #7440 and #7450, but I get the feeling most of the changes here will be either obsolete or should be backported to 6.0.x . I can continue working on this tomorrow.

joshcooper commented:
Putting blocked until merge ups happen.

DavidS commented:
Folded this into #7455 .

joshcooper commented:
Superseded by #7455
